### PR TITLE
Fix url_curl on MacOS

### PR DESCRIPTION
**Pull request by fanfuxiaoran**
Fix libcurl can not read data from gpfdist
on MacOS

But gpfdist with a pipe can not work on macos as
flock(2) which is used in gfile.c is not supported
on MacOS.

Fix the issue: https://github.com/greenplum-db/gpdb/issues/4105

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/access/external/url_curl.c [hunk 1]
```diff
@@ -686,14 +686,28 @@ fill_buffer(URL_CURL_FILE *curl, int want)
...
+		 * that isn't possible for your application to monitor with a socket and unfortunately
+		 * you can then not know exactly when the current action is completed using select().
+		 * You then need to wait a while before you proceed and call curl_multi_perform anyway
+		 */
+		if (maxfd == -1)
+		{
+			elog(DEBUG2, "curl_multi_fdset set maxfd = %d", maxfd);
+			pg_usleep(100);
+			// to call curl_multi_perform
+			nfds = 1;
```
Comment (lij55) [inline src/backend/access/external/url_curl.c:705]: is `continue` better than directly assignment here?

Comment (fanfuxiaoran) [reply src/backend/access/external/url_curl.c:705]: > is `continue` better than directly assignment here?

We can't call continue here. When maxfd is -1 means "curl_multi_fdset" can not extract fds from 
multi_handle. So it needs to call curl_multi_perform to create socket to connect to gpfdist. Only after the socket created,  curl_multi_fdset can extract maxfd from it.

I am also confused why there are no sockets connected to  gpfdist on MacOS  here, I need to hack more.

**Comment by pf-qiu** [comment]
Does this fix also work for https urls?

**Comment by fanfuxiaoran** [comment]
> Does this fix also work for https urls?

I thought it also works for https urls. But gpfdist_ssl test cases can not success on may Mac laptop,
but it looks like it's due to the lib of openssl (may be the version).

**Comment by pf-qiu** [comment]
I have a branch to fix the SSL here: https://github.com/pf-qiu/gpdb/tree/fix_ssl
I came to this accidentally, not sure about the cause. Can you check if this works for SSL on top of your patch?

**Comment by fanfuxiaoran** [comment]
> I have a branch to fix the SSL here: https://github.com/pf-qiu/gpdb/tree/fix_ssl
> I came to this accidentally, not sure about the cause. Can you check if this works for SSL on top of your patch?
Sure. But the gpfdist_ssl test cases are still failed.  And gpdb reported the following message in log:
" SSL certificate problem: self signed certificate"

**Review by huiliang-liu — APPROVED** [review]

File: src/backend/access/external/url_curl.c (hunk 1 above)
Comment (wuyuhao28) [reply src/backend/access/external/url_curl.c:705]: > > is `continue` better than directly assignment here?
> 
> We can't call continue here. When maxfd is -1 means "curl_multi_fdset" can not extract fds from multi_handle. So it needs to call curl_multi_perform to create socket to connect to gpfdist. Only after the socket created, curl_multi_fdset can extract maxfd from it.
> 
> I am also confused why there are no sockets connected to gpfdist on MacOS here, I need to hack more.

hi, I've met this problem on Linux too，did you find out why gpfdist has no socket ?

**Pull request merged**